### DISCUSSION TITLE
Registering sqlalchemy engine events

### DISCUSSION THREAD

**Discussion by TitaniumHocker** [pallets/flask#4785] [Q&A] [answered]
I need to register some event listeners on `sqlalchemy`'s engine(s). That can be done only if application context already appeared. There is `before_request` decorator to register callback for request, but I didn't found such callback decorator for application context.

Is there any way to run callback just after application context is created?

**Comment by davidism** [comment, marked as answer]
SQLAlchemy events are entirely part of SQLAlchemy. A Flask app context shouldn't be required to manage them. You should attach the events when creating the engine, not on every app context push.

Push an app context to get access to Flask-SQLAlchemy's engine. The same engine (pool) is reused across contexts, even though it requires an active context to actually get it.

```python
with app.app_context():
    engine = db.engine

    @sa.events.listen_for(engine, "event_name")
    def my_listener(...):
        ...
```

**Reply by TitaniumHocker** [reply]
I'm using `Flask-SQLAlchemy`, so engine objects only available in application context.
`flask_sqlalchemy.SQLAlchemy`'s engine property or `get_binds` function for getting engines for multiple binds is using `current_app` proxy, which is available only in application context.

**Reply by davidism** [reply]
Push an app context to get access to the engine. The same engine is reused across contexts, even though it requires an active context to actually get it.

```python
with app.app_context():
    engine = db.engine

    @sa.events.listen_for(engine, "event_name")
    def my_listener(...):
        ...
```

**Reply by TitaniumHocker** [reply]
If engines are reusing between contexts this should work. I'll try it, thx.

**Reply by TitaniumHocker** [reply]
Seems to work fine.
@davidism Can you copy your reply with pushing app context as new answer, so I can mark it as a reply?
